addon/mixins/legend.js

```diff
@@ -262,26 +284,121 @@ export default Ember.Mixin.create({
+          // and hence subtract short labels from remaining totalAvailableWidthForLabels.
+          totalAvailableWidthForLabels -= curLabelWidth;
+          // Distribute the remaining width equally in remaining labels and set that as max.
+          var remainingLabelCount = labelWidthsArray.length - (i + 1);
+          maxLabelWidth = totalAvailableWidthForLabels / remainingLabelCount;
+        }
+      }
+      return maxLabelWidth;
+    };
 
-    legendItems.append('text')
-          .style('text-anchor', 'start')
-          .text(function(d) { return d.label; })
-          .attr(this.get('legendLabelAttrs'))
-          .call(labelTrimmer.get('trim'));
+    // Select labels of current row (startIdx, endIdx) and truncate if greater than labelWidth
+    var truncateLabels = function(labels, startIdx, endIdx, labelWidth) {
+      var labelTrimmer = LabelTrimmer.create({
+        getLabelSize: function() {
+          return labelWidth;
+        }
+      });
+      // Select labels from current row and apply label trimmer
+      labels.filter(function(data, idx) {
+        return (idx >= startIdx && idx < endIdx);
+      }).call(labelTrimmer.get('trim'));
+    };
 
-    var totalLabelWidth = 0;
-    legend.selectAll('text').each(function() {
-      return totalLabelWidth += this.getComputedTextLength();
-    });
-    return this.set('averageLegendLabelWidth', totalLabelWidth / this.get('legendItems.length'));
+    // Select legendItems of current row (startIdx, endIdx) and calculate total row width
+    var calcLegendRowWidth = function(legendItems, startIdx, endIdx) {
+      var rowWidth = 0;
+      legendItems.filter(function(data, idx) {
+        return (idx >= startIdx && idx < endIdx);
+      }).each(function(val, col) {
+        rowWidth = (col === 0) ? 0 : (rowWidth + 2 * legendLabelPadding);
+        rowWidth += this.getBBox().width;
+      });
+      return rowWidth;
+    };
+
+    var startIdxCurRow = 0;
+    var legendRowWidths = []; // Capture the width of each legend row
+    var iconRadius = this.get('legendIconRadius');
+    var iconToLabelPadding = iconRadius * 0.5 + legendLabelPadding;
+    var legendItemPadding = 2 * legendLabelPadding;
+
+    // Perform label truncation for legend items in each row.
+    for (rowNum = 0; rowNum < numLegendItemsByRows.length; rowNum++) {
+      var curRowItemCount = numLegendItemsByRows[rowNum];
+      // Legend layout => [(-legendIconRadius, 0) "Legend Icon" (legendIconRadius, 0)] (legendIconRadius/2 + legendLabelPadding) "Label" (2*legendLabelPadding) [Next Legend in the row]
+      var totalAvailableWidthForLabels = legendWidth - // Total width of a legend row available in the chart
+                                curRowItemCount * (iconRadius + iconToLabelPadding) - // Subtract width of each icon and it's padding
+                                (curRowItemCount - 1) * legendItemPadding; // Subtract width of all padding between items
```

Comment:
Pull each of these expressions out into temporary variables with names that communicate what the comments are doing.